symfit/core/fit.py

```diff
@@ -1235,93 +1027,147 @@ def execute(self, relative_error=1e-8, max_iter=500):
+            parameters and returns a scalar.
+        :param minimizer: Have Fit use your specified minimizer.
+        :param ordered_data: data for dependent, independent and sigma variables. Assigned in
+            the following order: independent vars are assigned first, then dependent
+            vars, then sigma's in dependent vars. Within each group they are assigned in
+            alphabetical order.
+        :param named_data: assign dependent, independent and sigma variables data by name.
         """
-        constraints = kwargs.pop('constraints')
-        super(Minimize, self).__init__(model, *args, **kwargs)
+        objective = named_data.pop('objective')
+        minimizer = named_data.pop('minimizer')
+        constraints = named_data.pop('constraints')
+        super(Fit, self).__init__(model, *ordered_data, **named_data)
+
+        # List of Constraint objects
+        self.constraints = self._init_constraints(constraints=constraints)
+
+        if objective is None:
+            # Param only scalar Model -> the model is the objective.
+            if len(self.model.independent_vars) == 0 and len(self.model) == 1:
+                # No data provided means a simple minimization of the Model parameters
+                # is requested, not a fit.
+                if all(value is None for value in self.data.values()):
+                    objective = MinimizeModel
+            elif minimizer is MINPACK:
+                # MINPACK is considered a special snowflake, as its API has to be
+                # considered seperately and has its own non standard objective function.
+                objective = VectorLeastSquares
+
+        if objective is None:
+            objective = LeastSquares
+
+        # Initialise the objective if it's not initialised already
+        if isinstance(objective, BaseObjective):
+            self.objective = objective
+        else:
+            self.objective = objective(self.model, self.data)
+
+        # Select the minimizer on the basis of the provided information.
+        if minimizer is None:
+            if self.constraints:
+                minimizer = SLSQP
+            elif any([bound is not None for pair in self.model.bounds for bound in pair]):
+                # If any bound is set
+                minimizer = LBFGSB
+            else:
+                minimizer = BFGS
 
-        self.constraints = []
+        # Initialise the minimizer
+        if isinstance(minimizer, BaseMinimizer):
+            self.minimizer = minimizer
+        else:
+            minimizer_options = {}
+            if issubclass(minimizer, GradientMinimizer):
+                # if isinstance(self.model, ODEModel):
+                #     raise Exception(self.model.numerical_jacobian)
+                if hasattr(self.model, 'numerical_jacobian') and hasattr(self.objective, 'eval_jacobian'):
+                    minimizer_options['jacobian'] = self.objective.eval_jacobian
```

Comment:
Yes, because if no jacobian is provided explicitly it will be estimated numerically. So only if the model provides an analytical jacobian should `eval_jacobian` be preferred over letting the minimizer work it's magic. Actually, this goes back nicely to our discussion that `eval_jacobian` calls with finite differences are relatively expencive. I think that is why you choose to check for both `eval_jacobian` and `numerical_jacobian`, the latter indicating if it has an analytical form.

I will add all this as a comment to the code though.